docs/tutorials/writing-views.md

```diff
@@ -147,58 +149,109 @@ Here's an example:
 realm administrator (and thus has permission to create a user); we
-show it here primarily to show how `has_request_variables` should be
+show it here primarily to show how `typed_endpoint` should be
 the inner decorator.
 
-The implementation of `has_request_variables` is documented in detail
+The implementation of `typed_endpoint` is documented in detail
 in
-[zerver/lib/request.py](https://github.com/zulip/zulip/blob/main/zerver/lib/request.py))
+[zerver/lib/typed_endpoint.py](https://github.com/zulip/zulip/blob/main/zerver/lib/typed_endpoint.py)
 
-REQ also helps us with request variable validation. For example:
+Pydantic also helps us with request variable validation. For example:
 
-- `msg_ids = REQ(json_validator=check_list(check_int))` will check
-  that the `msg_ids` HTTP parameter is a list of integers, marshalled
-  as JSON, and pass it into the function as the `msg_ids` Python
+- `msg_ids: Json[list[int]]` will check that the `msg_ids`
+  HTTP parameter is a list of integers, marshalled as JSON,
+  and pass it into the function as the `msg_ids` Python
   keyword argument.
 
-- `streams_raw = REQ("subscriptions", json_validator=check_list(check_string))`
+- `streams_raw: Annotated[Json[list[str]], ApiParamConfig("subscriptions")]`
   will check that the "subscriptions" HTTP parameter is a list of
   strings, marshalled as JSON, and pass it into the function with the
   Python keyword argument `streams_raw`.
 
-- `message_id=REQ(converter=to_non_negative_int)` will check that the
-  `message_id` HTTP parameter is a string containing a non-negative
-  integer (`converter` differs from `json_validator` in that it does
-  not automatically marshall the input from JSON).
+- `message_id: Json[NonNegativeInt]` will check that the `message_id`
+  HTTP parameter is a string containing a JSON encoded non-negative
+  integer.
 
-- Since there is no need to JSON-encode strings, usually simply
-  `my_string=REQ()` is correct. One can pass, for example,
-  `str_validator=check_string_in(...)` where one wants to run a
+[Annotated](https://docs.python.org/3/library/typing.html#typing.Annotated)
+can be used in combination with
+[Pydantic's validators](https://docs.pydantic.dev/latest/api/functional_validators/)
+to provide additional validation for the arguments.
+
+- `name: Annotated[str, StringConstraints(max_length=60)]` will check that the
+  `name` HTTP parameter is a string containing up to 60 characters.
+
+- Since there is no need to JSON-encode strings
+  (lists, integers, bools and complex objects require JSON encoding), usually simply
+  `my_string: str` is correct. One can pass, for example,
+  `Annotated[str, check_string_in_validator(...)]` where one wants to run a
   validator on the value of a string.
 
+Default values can be specified for optional arguments similar to how we would specify
+default values in regular python function.
+
+- `mandatory: Json[bool] = False` will assign False to the `mandatory` argument if
```

Comment:
`mandatory` as the parameter name feels like a juxtaposition here. Maybe something more generic?